src/main.rs

```diff
@@ -1,245 +1,202 @@
-    marker::Unpin,
-    str::FromStr,
+    io::{self, Cursor, Write},
+    str::{self, FromStr},
 };
 use url::Url;
 use wasmer::{imports, Function, Instance, Module, Store, Type, Value};
 
-/// Main entry point.
 #[tokio::main]
-async fn main() -> Result<(), Box<dyn Error>> {
+async fn main() -> Result<()> {
     env_logger::init();
 
-    let (mut network_client, mut network_events, network_event_loop) = network::new(None).await?;
-
-    spawn(network_event_loop.run());
-
     let opts = Args::parse();
+    let keypair = Keypair::generate_ed25519();
+    let swarm = swarm::build_swarm(keypair).await?;
+    let (mut client, mut events, event_loop) = Client::new(swarm).await?;
+
+    tokio::spawn(event_loop.run());
 
     if let Some(addr) = opts.peer {
         let peer_id = match addr.iter().last() {
             Some(Protocol::P2p(hash)) => PeerId::from_multihash(hash).expect("Valid hash."),
-            _ => return Err("Expect peer multiaddr to contain peer ID.".into()),
+            _ => bail!("Expect peer multiaddr to contain peer ID."),
         };
-        network_client
-            .dial(peer_id, addr)
-            .await
-            .expect("Dial to succeed");
+        client.dial(peer_id, addr).await.expect("Dial to succeed.");
     }
 
     match opts.listen {
-        Some(addr) => network_client
+        Some(addr) => client
             .start_listening(addr)
             .await
             .expect("Listening not to fail."),
 
-        None => network_client
+        None => client
             .start_listening("/ip4/0.0.0.0/tcp/0".parse()?)
             .await
             .expect("Listening not to fail."),
     };
 
     match opts.argument {
-        Argument::Get { name } => request(name, &mut network_client).await,
-        Argument::Provide { wasm, fun, args } => {
-            provide(wasm, fun, args, &mut network_client, &mut network_events).await
-        }
-    }
-
-    Ok(())
-}
-
-async fn request(name: String, network_client: &mut network::Client) -> () {
-    let providers = network_client.get_providers(name.clone()).await;
-    if providers.is_empty() {
-        panic!("Could not find provider for file {}.", name);
-    }
-
-    let requests = providers.into_iter().map(|p| {
-        let mut network_client = network_client.clone();
-        let name = name.clone();
-        async move { network_client.request_file(p, name).await }.boxed()
-    });
-
-    let file_map = futures::future::select_ok(requests)
-        .await
-        .map_err(|_| "None of the providers returned file.");
-
-    let file_content = match file_map {
-        Ok(x) => x.0,
-        Err(y) => panic!("no file {}", y),
-    };
+        Argument::Get { name } => {
+            let providers = client.get_providers(name.clone()).await?;
+            let providers = providers
+                .is_empty()
+                .then_some(providers)
+                .ok_or_else(|| anyhow!("Could not find provider for file {name}."))?;
```

Comment:
Why the reassignment? Can this not be...

```rust
let providers =
  client
    .get_providers(name.clone())
    .await?
    .is_empty()
    .then_some(providers)
    .ok_or_else(|| anyhow!("Could not find provider for file {name}."))?;
```

...or is that unidiomatic?

Reply:
there was a reason during cleanup, but I think you're right that it flows through :). I'll do this now.

Reply:
I certainly have a lot of dangling stuff like this in my current code as well! Not a criticism or anything; just letting me inner 5 year old out. "Why? Why? Why? Why?" 😁

Reply:
@expede oh yeah, this won't work. That's due to how then_some expects the input.